signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,314 @@
+        Ok(libp2p::swarm::dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: libp2p::core::Endpoint,
+        _port_use: libp2p::core::transport::PortUse,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        Ok(libp2p::swarm::dummy::ConnectionHandler)
+    }
+
+    fn handle_pending_outbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _maybe_peer: Option<PeerId>,
+        addresses: &[Multiaddr],
+        effective_role: libp2p::core::Endpoint,
+    ) -> Result<Vec<Multiaddr>, libp2p::swarm::ConnectionDenied> {
+        tracing::debug!(?addresses, "handling pending outbound connection");
+        let addresses = addresses.to_vec();
+
+        // We're only interested in our outbound dialing activity. This _can_ be
+        // a listener if the swarm is attempting a hole-punch.
+        if !effective_role.is_dialer() {
+            return Ok(addresses);
+        }
+
+        for addr in &addresses {
+            if self.is_seed_address(addr) {
+                tracing::debug!(%addr, "attempting to dial seed address");
+            }
+        }
+
+        // Return the addresses as-is.
+        Ok(addresses)
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm) {
+        match event {
+            FromSwarm::ConnectionEstablished(e) => {
+                // We're only interested in our outbound activity.
+                if !e.endpoint.is_dialer() {
+                    return;
+                }
+
+                // If we've connected to ourselves, we don't need to do anything.
+                // This might happen if a node has its own address in the seed list.
+                if e.peer_id == self.local_peer_id() {
+                    return;
+                }
+
+                // Increment our total connection count. We're not interested
+                // in counting connections to ourselves.
+                self.connection_count = self.connection_count.saturating_add(1);
+
+                // Get the remote address, and remove the p2p protocol if it exists.
+                let mut addr = e.endpoint.get_remote_address().clone();
+                if matches!(addr.iter().last(), Some(Protocol::P2p(_))) {
+                    addr.pop();
+                }
+
+                // Check if the address is a seed address. If it's not, then
+                // we don't need to do anything so we return early.
+                if self.is_seed_address(&addr) {
+                    tracing::debug!(peer_id = %e.peer_id, %addr, "connected to seed");
+                } else {
+                    return;
+                }
+
+                // Update our connected seeds map.
+                match self.connected_seeds.entry(e.peer_id) {
```

Comment:
nit: we could simplify this match with something like `self.connected_seeds.entry(e.peer_id).or_default().insert(addr);` (eventually with the check and log if inserted).

Reply:
Yup, good one [76f0815](https://github.com/stacks-network/sbtc/pull/1257/commits/76f08156e6574538cc336264ed22c1d96769d094)